core/src/main/scala/latis/model/ModelParser.scala

```
@@ -5,49 +5,62 @@ import Atto._
+    ModelParser.parse(exp)
 
   private def variable: Parser[String] =
     sepBy1(identifier, char('.')).map(_.toList.mkString("."))
 
-  private def scalar: Parser[Scalar] = variable.map { id =>
-    Scalar(Metadata(
-      "id" -> id,
-      "type" -> "double" //TODO: allow in spec
-    ))
-  }
-
-  private def tuple: Parser[Tuple] = for {
-    _ <- char('(')
-    vars <- many(scalar)
-    _ <- many(whitespace) ~ char(',') ~ many(whitespace)
-    last <- scalar
-    _ <- char(')')
-  } yield Tuple(vars :+ last)
-
-  private def function: Parser[Function] = for {
-    domain <- scalar | tuple
-    _ <- many(whitespace) ~ string("->") ~ many(whitespace)
-    range <- dataType
-  } yield Function(domain, range)
-
   private def dataType: Parser[DataType] =
     function | tuple | scalar
+
+  /** Doesn't support nested functions in the domain. */
+  def function: Parser[DataType] =
+    parens(functionWithoutParens) | functionWithoutParens
+
+  /** Doesn't support nested functions in the domain. */
+  private def functionWithoutParens: Parser[DataType] = for {
+    d <- (tuple | scalar).token
+    _ <- string("->").token
+    r <- dataType.token
+  } yield Function(d, r)
+
+  /** Only parses tuples of scalars */
+  def tuple: Parser[DataType] = for {
+    l <- parens(sepBy(scalar.token, char(',').token))
+  } yield Tuple(l)
+
+  def scalar: Parser[DataType] =
+    scalarWithType | scalarWithoutType
+
+  private def scalarWithType: Parser[DataType] = for {
+    id <- variable.token
+    _ <- string(":").token
+    t <- valueType.token
+  } yield Scalar(Metadata(id) + ("type" -> t.toString))
+
+  private def scalarWithoutType: Parser[DataType] = for {
+    id <- variable.token
+  } yield Scalar(Metadata(id) + ("type" -> defaultScalarType))
+
+  private def valueType: Parser[ValueType] = for {
+    tName <- many(letter)
+    tString = tName.mkString("")
+  } yield ValueType.fromName(tString).fold(throw _, identity)
```

Can we raise an invalid value type into a parse failure?

---

This will throw an exception:

```
@ ModelParser("a: fake -> b: fake") 
latis.util.LatisException: Invalid Scalar value type: fake
```

I'd prefer that this fail in `Parser` rather than throw an exception.